From: Alessandro Astone <alessandro.astone@canonical.com>
Date: Fri, 19 Sep 2025 13:32:51 +0200
Subject: xi2: Do not discard emulated pointer events during drag-and-drop

The filter for XIPointerEmulated XI_Motion events, which was meant to
filter-out emulated scroll events, was also inadvertently filtering-out
the emulated drag-and-drop motion updates.

Forwarded: https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/8951
Bug-Ubuntu: https://launchpad.net/bugs/2125309
---
 gdk/x11/gdkdevicemanager-xi2.c | 19 +++++++++----------
 1 file changed, 9 insertions(+), 10 deletions(-)

diff --git a/gdk/x11/gdkdevicemanager-xi2.c b/gdk/x11/gdkdevicemanager-xi2.c
index 0fc73f3..086995b 100644
--- a/gdk/x11/gdkdevicemanager-xi2.c
+++ b/gdk/x11/gdkdevicemanager-xi2.c
@@ -1621,11 +1621,6 @@ gdk_x11_device_manager_xi2_translate_event (GdkEventTranslator *translator,
                            xev->detail,
                            xev->event_x, xev->event_y);
 
-#ifdef XINPUT_2_2
-        if (xev->flags & XIPointerEmulated)
-          return FALSE;
-#endif
-
         if (ev->evtype == XI_ButtonRelease &&
             (xev->detail >= 4 && xev->detail <= 7))
           return FALSE;
@@ -1636,6 +1631,10 @@ gdk_x11_device_manager_xi2_translate_event (GdkEventTranslator *translator,
 
             /* Button presses of button 4-7 are scroll events */
 
+#ifdef XINPUT_2_2
+        if (xev->flags & XIPointerEmulated)
+          return FALSE;
+#endif
             if (xev->detail == 4)
               direction = GDK_SCROLL_UP;
             else if (xev->detail == 5)
@@ -1706,11 +1705,6 @@ gdk_x11_device_manager_xi2_translate_event (GdkEventTranslator *translator,
         double x, y;
         double *axes;
 
-#ifdef XINPUT_2_2
-        if (xev->flags & XIPointerEmulated)
-          return FALSE;
-#endif
-
         source_device = g_hash_table_lookup (device_manager->id_table,
                                              GUINT_TO_POINTER (xev->sourceid));
         device = g_hash_table_lookup (device_manager->id_table,
@@ -1732,6 +1726,11 @@ gdk_x11_device_manager_xi2_translate_event (GdkEventTranslator *translator,
             GdkModifierType state;
             GdkScrollDirection direction;
 
+#ifdef XINPUT_2_2
+        if (xev->flags & XIPointerEmulated)
+          return FALSE;
+#endif
+
             GDK_DISPLAY_DEBUG (display, EVENTS,
                                "smooth scroll: \n\tdevice: %u\n\tsource device: %u\n\twindow %ld\n\tdeltas: %f %f",
                                 xev->deviceid, xev->sourceid,
